src/parse.rs

```diff
@@ -817,180 +793,181 @@ impl<'a> Bytes<'a> {
-                    Some(b'#' | b'"') => return Err(Error::ExpectedIdentifier),
-                    Some(_) | None => (),
-                },
-                Some(_) | None => (),
-            }
-        };
+        #[allow(clippy::nonminimal_bool)]
+        if self.check_str("b\"") // byte string
+            || self.check_str("b'") // byte literal
+            || self.check_str("br#") // raw byte string
+            || self.check_str("br\"") // raw byte string
+            || self.check_str("r\"") // raw string
+            || self.check_str("r#\"") // raw string
+            || self.check_str("r##") // raw string
+            || false
+        {
+            return Err(Error::ExpectedIdentifier);
+        }
 
-        let length = if next == b'r' {
-            match self.bytes.get(1) {
-                Some(b'"') => return Err(Error::ExpectedIdentifier),
-                Some(b'#') => {
-                    let after_next = self.bytes.get(2).copied().unwrap_or_default();
-                    // Note: it's important to check this before advancing forward, so that
-                    // the value-type deserializer can fall back to parsing it differently.
-                    if !is_ident_raw_char(after_next) {
-                        return Err(Error::ExpectedIdentifier);
-                    }
-                    // skip "r#"
-                    let _ = self.advance(2);
-                    self.next_bytes_contained_in(is_ident_raw_char)
-                }
-                _ => {
-                    let std_ident_length = self.next_bytes_contained_in(is_ident_other_char);
-                    let raw_ident_length = self.next_bytes_contained_in(is_ident_raw_char);
+        let length = if self.check_str("r#") {
+            let cursor_backup = self.cursor;
 
-                    if raw_ident_length > std_ident_length {
-                        if let Ok(ident) = std::str::from_utf8(&self.bytes[..raw_ident_length]) {
-                            return Err(Error::SuggestRawIdentifier(String::from(ident)));
-                        }
-                    }
+            self.advance(2);
 
-                    std_ident_length
-                }
+            // Note: it's important to check this before advancing forward, so that
+            // the value-type deserializer can fall back to parsing it differently.
+            if !matches!(self.peek(), Some(c) if is_ident_raw_char(c)) {
+                self.set_cursor(cursor_backup);
+                return Err(Error::ExpectedIdentifier);
+            }
+
+            self.next_chars_while(is_ident_raw_char)
+        } else if first == 'r' {
+            let std_ident_length = self.next_chars_while(is_xid_continue);
+            let raw_ident_length = self.next_chars_while(is_ident_raw_char);
+
+            if raw_ident_length > std_ident_length {
+                return Err(Error::SuggestRawIdentifier(
+                    self.src()[1..raw_ident_length].into(),
```

Comment:
Should this really be offset by one? Maybe the leading r is part of the identifier?